### ISSUE TITLE
Build acceleration might not work with slow cheetah

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/project-system#8886]
Reported by @Dunge in https://github.com/dotnet/project-system/issues/8804#issuecomment-1448933862

> I'm using [slow-cheetah](https://github.com/microsoft/slow-cheetah) as a nuget to transform .config files based on the build configuration. While my projects themselves don't have post build steps, I believe this tool create one internally. Took me a while to understand why some config files would _sometimes_ not get transformed properly, until I realized this happens only since I enabled Visual Studio Build Acceleration.

We should investigate this. It might be a problem with slow cheetah, or with build acceleration. Either way we should understand why build acceleration causes this to start failing and do our best to prevent this kind of error.

**Comment by drewnoakes** [comment]
@Dunge are you able to provide a repro, or at least describe the scenario in more detail? If you can share verbose log output from the up-to-date check during build, that would also be helpful.

I would expect the project that produces the transformations to correctly advertise its inputs/outputs to the fast up-to-date check such that we know to build it. It may be that slow cheetah needs to extend its `.targets` files (such as [this one](https://github.com/microsoft/slow-cheetah/blob/master/src/Microsoft.VisualStudio.SlowCheetah/Build/Microsoft.VisualStudio.SlowCheetah.targets)).

**Comment by Dunge** [comment]
@drewnoakes 

Sure! Here is a repro solution: [Dunge/FastBuildTransform](https://github.com/Dunge/FastBuildTransform)

I based it on my real project. A NLog.config file that points to a null logging, and a NLog.Debug.config transformation file that make the output points towards console logging.

Took me a while to understand how to replicate the issue, but it occurs when referencing a class library and you edit the code of this class library, but not the main app.

Build and launch the app once, you should see the `Worker console logging - change me` message.
Edit the string under the  `ClassLibrary\Worker.cs` file, build and launch. The console will remains empty, and the NLog.config file in the output directory won't have been transformed (points to null target).
Edit the string under the  `FastBuildTransform\Program.cs` file (non referenced variable), build and launch. The console will work again with transformation applied.

Do the same test without the `AccelerateBuildsInVisualStudio` in `Directory.build.props` and it will get transformed in all situations.

**Comment by drewnoakes** [comment]
Thanks so much for the repro! It's great to have something concrete to refer to.

<details>
<summary>Clean and <kbd>F5</kbd> produces build log (app writes to console)</summary>

```
1>FastUpToDate: Ignoring up-to-date check items with Kind="ImplicitBuild" (ClassLibrary)
1>FastUpToDate: Comparing timestamps of inputs and outputs: (ClassLibrary)
1>FastUpToDate:     Adding UpToDateCheckBuilt outputs: (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll (ClassLibrary)
1>FastUpToDate: Output 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll' does not exist, not up-to-date. (ClassLibrary)
1>FastUpToDate: Up-to-date check completed in 0.5 ms (ClassLibrary)
1>------ Build started: Project: ClassLibrary, Configuration: Debug Any CPU ------
Restored D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\FastBuildTransform.csproj (in 68 ms).
Restored D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\ClassLibrary.csproj (in 68 ms).
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>ClassLibrary -> D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll
2>FastUpToDate: Ignoring up-to-date check items with Kind="ImplicitBuild" (FastBuildTransform)
2>FastUpToDate: Comparing timestamps of inputs and outputs: (FastBuildTransform)
2>FastUpToDate:     Adding UpToDateCheckBuilt outputs: (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\FastBuildTransform.dll (FastBuildTransform)
2>FastUpToDate: Output 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\FastBuildTransform.dll' does not exist, not up-to-date. (FastBuildTransform)
2>FastUpToDate: Up-to-date check completed in 0.6 ms (FastBuildTransform)
2>------ Build started: Project: FastBuildTransform, Configuration: Debug Any CPU ------
2>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
2>FastBuildTransform -> D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\FastBuildTransform.dll
========== Build: 2 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

</details>

<details>
<summary>Change `Worker.cs` and <kbd>F5</kbd> produces build log (app doesn't write to console)</summary>

```
Build started...
1>FastUpToDate: Ignoring up-to-date check items with Kind="ImplicitBuild" (ClassLibrary)
1>FastUpToDate: Comparing timestamps of inputs and outputs: (ClassLibrary)
1>FastUpToDate:     Adding UpToDateCheckBuilt outputs: (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\obj\Debug\net6.0\ClassLibrary.dll (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\obj\Debug\net6.0\ClassLibrary.pdb (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.pdb (ClassLibrary)
1>FastUpToDate:     Adding project file inputs: (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\ClassLibrary.csproj (ClassLibrary)
1>FastUpToDate:     Adding newest import input: (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\obj\ClassLibrary.csproj.nuget.g.targets (ClassLibrary)
1>FastUpToDate:     Adding Compile inputs: (ClassLibrary)
1>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\Worker.cs (ClassLibrary)
1>FastUpToDate: Input Compile item 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\Worker.cs' is newer (2023-03-01 18:15:43.345) than earliest output 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\obj\Debug\net6.0\ClassLibrary.pdb' (2023-03-01 18:14:39.968), not up-to-date. (ClassLibrary)
1>FastUpToDate: Up-to-date check completed in 1.6 ms (ClassLibrary)
1>------ Build started: Project: ClassLibrary, Configuration: Debug Any CPU ------
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>ClassLibrary -> D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll
2>FastUpToDate: Ignoring up-to-date check items with Kind="ImplicitBuild" (FastBuildTransform)
2>FastUpToDate: Comparing timestamps of inputs and outputs: (FastBuildTransform)
2>FastUpToDate:     Adding UpToDateCheckBuilt outputs: (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\FastBuildTransform.dll (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\obj\Debug\net6.0\FastBuildTransform.dll (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\obj\Debug\net6.0\FastBuildTransform.pdb (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\FastBuildTransform.pdb (FastBuildTransform)
2>FastUpToDate:     Adding project file inputs: (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\FastBuildTransform.csproj (FastBuildTransform)
2>FastUpToDate:     Adding newest import input: (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\obj\FastBuildTransform.csproj.nuget.g.targets (FastBuildTransform)
2>FastUpToDate:     Adding Compile inputs: (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\Program.cs (FastBuildTransform)
2>FastUpToDate:     Adding ResolvedCompilationReference inputs: (FastBuildTransform)
2>FastUpToDate:         D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\obj\Debug\net6.0\ref\ClassLibrary.dll (FastBuildTransform)
2>FastUpToDate:     No inputs are newer than earliest output 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\obj\Debug\net6.0\FastBuildTransform.pdb' (2023-03-01 18:14:40.533). Newest input is 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\obj\Debug\net6.0\ref\ClassLibrary.dll' (2023-03-01 18:14:39.972). (FastBuildTransform)
2>FastUpToDate: Checking items to copy to the output directory: (FastBuildTransform)
2>FastUpToDate:     Checking copy items from project 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\FastBuildTransform.csproj': (FastBuildTransform)
2>FastUpToDate:         Checking Always item (FastBuildTransform)
2>FastUpToDate:             Source      2023-03-01 17:59:23.155: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\NLog.config' (FastBuildTransform)
2>FastUpToDate:             Destination 2023-03-01 18:14:40.600: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\NLog.config' (FastBuildTransform)
2>FastUpToDate:             Remembering the need to copy file 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\NLog.config' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\NLog.config'. (FastBuildTransform)
2>FastUpToDate:         Checking PreserveNewest item (FastBuildTransform)
2>FastUpToDate:             Source      2023-03-01 18:14:40.553: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\obj\Debug\net6.0\apphost.exe' (FastBuildTransform)
2>FastUpToDate:             Destination 2023-03-01 18:14:40.553: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\FastBuildTransform.exe' (FastBuildTransform)
2>FastUpToDate:     Checking copy items from project 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\ClassLibrary.csproj': (FastBuildTransform)
2>FastUpToDate:         Checking PreserveNewest item (FastBuildTransform)
2>FastUpToDate:             Source      2023-03-01 18:15:44.156: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll' (FastBuildTransform)
2>FastUpToDate:             Destination 2023-03-01 18:14:39.971: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.dll' (FastBuildTransform)
2>FastUpToDate:             Remembering the need to copy file 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.dll'. (FastBuildTransform)
2>FastUpToDate:         Checking PreserveNewest item (FastBuildTransform)
2>FastUpToDate:             Source      2023-03-01 18:15:44.154: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.pdb' (FastBuildTransform)
2>FastUpToDate:             Destination 2023-03-01 18:14:39.968: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.pdb' (FastBuildTransform)
2>FastUpToDate:             Remembering the need to copy file 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.pdb' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.pdb'. (FastBuildTransform)
2>FastUpToDate: Copying 3 files to accelerate build (https://aka.ms/vs-build-acceleration): (FastBuildTransform)
2>FastUpToDate:     From 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\NLog.config' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\NLog.config'. (FastBuildTransform)
2>FastUpToDate:     From 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.dll'. (FastBuildTransform)
2>FastUpToDate:     From 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.pdb' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.pdb'. (FastBuildTransform)
2>FastUpToDate: Build acceleration copied 3 files. (FastBuildTransform)
2>FastUpToDate: Project is up-to-date. (FastBuildTransform)
2>FastUpToDate: Up-to-date check completed in 9.7 ms (FastBuildTransform)
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
Visual Studio accelerated 1 project(s), copying 3 file(s). See https://aka.ms/vs-build-acceleration.
```

</details>

The key lines in the second log are:

```
2>FastUpToDate: Checking items to copy to the output directory: (FastBuildTransform)
2>FastUpToDate:     Checking copy items from project 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\FastBuildTransform.csproj': (FastBuildTransform)
2>FastUpToDate:         Checking Always item (FastBuildTransform)
2>FastUpToDate:             Source      2023-03-01 17:59:23.155: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\NLog.config' (FastBuildTransform)
2>FastUpToDate:             Destination 2023-03-01 18:14:40.600: 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\NLog.config' (FastBuildTransform)
2>FastUpToDate:             Remembering the need to copy file 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\NLog.config' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\NLog.config'. (FastBuildTransform)
...
2>FastUpToDate: Copying 3 files to accelerate build (https://aka.ms/vs-build-acceleration): (FastBuildTransform)
2>FastUpToDate:     From 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\NLog.config' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\NLog.config'. (FastBuildTransform)
2>FastUpToDate:     From 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.dll' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.dll'. (FastBuildTransform)
2>FastUpToDate:     From 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\ClassLibrary\bin\Debug\net6.0\ClassLibrary.pdb' to 'D:\repos\issues\Issue8886BuildAccelerationAndTransformedCode\FastBuildTransform\FastBuildTransform\bin\Debug\net6.0\ClassLibrary.pdb'. (FastBuildTransform)
```

The coping of `ClassLibrary.pdb` and `ClassLibrary.dll` are expected. However the copying of `NLog.config` is undoing the work done by the last build's transformation.

VS has been told that the `NLog.config` file needs must _Always_ be copied to the output directory, so is doing so.

Here's why:

```xml
  <ItemGroup>
    <None Include="NLog.config">
      <SubType>Designer</SubType>
      <TransformOnBuild>true</TransformOnBuild>
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
    <None Include="NLog.Debug.config">
      <IsTransformFile>True</IsTransformFile>
      <DependentUpon>NLog.config</DependentUpon>
    </None>
  </ItemGroup>
```

Changing the `CopyToOutputDirectory` value from `Always` to `PreserveNewest` fixes your issue.

**Comment by drewnoakes** [comment]
One option here is to disable Build Acceleration when we detect a Slow Cheetah NuGet package is present on the project. We added a facility for this in #9214.

Another option is to hard-code knowledge of Slow Cheetah into the project system. I'm less inclined to do this however.

A [code search on GitHub](https://github.com/search?type=code&q=%3CTransformOnBuild%3Etrue%3C%2FTransformOnBuild%3E+language%3AXML) shows the problem you encountered would be relatively widespread.

I filed #9258 to disable Build Acceleration when Slow Cheetah is present.
